opensrp-app/src/main/java/org/smartregister/util/FormUtils.java

```diff
@@ -1146,25 +1204,60 @@ public JSONObject getFormJsonFromRepositoryOrAssets(String formIdentity) {
             String revisedFormName = extractFormNameWithoutExtension(localeFormIdentity);
             clientForm = clientFormRepository.getActiveClientFormByIdentifier(revisedFormName);
         }
+        return clientForm;
+    }
 
-        try {
-            if (clientForm != null) {
-                Timber.d("============%s form loaded from db============", formIdentity);
+    public void handleJsonFormOrRulesError(@NonNull String formIdentity, @NonNull OnFormFetchedCallback<String> onFormFetchedCallback) {
+        handleJsonFormOrRulesError(false, formIdentity, onFormFetchedCallback);
+    }
 
-                JSONObject formJson = new JSONObject(clientForm.getJson());
-                injectFormStatus(formJson, clientForm);
+    public void handleJsonFormOrRulesError(boolean isRulesFile, @NonNull String formIdentity, @NonNull OnFormFetchedCallback<String> onFormFetchedCallback) {
+        ClientForm clientForm = getClientFormFromRepository(formIdentity);
+        ClientFormRepository clientFormRepository = CoreLibrary.getInstance().context().getClientFormRepository();
+        List<ClientForm> clientForms = clientFormRepository.getClientFormByIdentifier(clientForm.getIdentifier());
+
+        if (clientForms.size() > 0) {
+            // Show dialog asking user if they want to rollback to the previous available version X
+            // if YES, then provide that form instead
```

Comment:
Are we checking for the APK compatibility when suggesting the form rollback version.

Reply:
Nope, we are not. As discussed, we provide the user with all the available forms. 

1. The logic for providing the option ended up being too complicated & would take a lot of effort
2. Our assumption that app version means a change in app-model is not accurate. The idea is that the model might remain the same across multiple app versions and thus we cannot accurately determine or suggest to the user the model-compatible forms

Reply:
What happens if I update the registration form & update the registration form logic and add/remove new questions or hidden fields e.g `contact_no` which are key to the rest of the app logic? If we roll back to a version that doesn't have this field then use the new APK then the logic will be messed up.

Reply:
We do not automatically rollback, we give the user the option, We cannot also tell critical and non-critical changes. That would require extra work to determine. What do you think?Do you have a suggestion that accurately determines compatible rollback form(s)?